PocketKit/Sources/PocketKit/Listen/ListenViewModel.swift

```diff
@@ -19,19 +36,80 @@ class ListenViewModel: PKTListenDataSource<PKTListDiffable> {
             .compactMap { PKTListenKusariCreate($0.albumID!, PKTListenQueueSectionType.item.rawValue, $0, config) }
         ?? []
 
-        DispatchQueue.global(qos: .background).async {
-            // Warm up the first 6 images
-            listenItems.prefix(6).forEach({ listenItem in
-                listenItem.warmImage()
-            })
+        defer {
+            DispatchQueue.global(qos: .background).async {
+                // Warm up the first 6 images
+                listenItems.prefix(6).forEach({ listenItem in
+                    listenItem.warmImage()
+                })
+            }
         }
 
-        let viewModel = ListenViewModel(context: ["index": NSNumber(value: 0)]) { source, context, complete in
+        self.init(context: ["index": NSNumber(value: 0)]) { source, context, complete in
             source.hasMore = false
             Log.debug("Loaded Listen with \(listenItems.count) articles")
             complete(nil, ["index": NSNumber(value: listenItems.count)], listenItems)
         }
-        viewModel.title = title
-        return viewModel
+
+        self.savedItems = savedItems
+    }
+
+    func textUnits(for kusari: PKTKusari<PKTListenItem>) -> [PKTTextUnit] {
+        // When returning the text components for offline TTS,
+        // fetch the saved item with the given URL of the item staged for playing, ensuring it has an article
+        guard let givenURL = kusari.album?.givenURL,
+              let savedItem = savedItems?.first(where: { $0.givenURL == givenURL }),
+              let article = savedItem.item?.article
+        else {
+            return []
+        }
+
+        // For offline TTS, filter out only components that contain text "blocks", returning the appropriate types
+        return article.components.compactMap { c in
+            switch c {
+            case .blockquote(let component): return component.content
+            case .bulletedList(let component): return string(from: component)
```

Comment:
I wonder if things like blocqutes and lists should have text added `List`, etc.

Reply:
I kind of want to follow the K.I.S.S method for this - start simple, and iterate as we think articles can be improved, or receive feedback of such. I'm not sure what the legacy app did, but I don't imagine it read out the type of incoming block of text. Rather, it just treated it _as text_ (but that's an assumption).